From a6c8149af3aa2094fcaceaa38fb333a2f10f2a98 Mon Sep 17 00:00:00 2001
From: Christian Marangi <ansuelsmth@gmail.com>
Date: Mon, 24 Oct 2022 22:13:30 +0200
Subject: [PATCH] sh: add CONFIG_CMDLINE_OVERRIDE option

Some bootloaders may have uneditable bootargs. While CMDLINE_FORCE can
be used, this is not a good option for kernels that are shared across
devices. This setting enables using "chosen/cmdline-override" binding
as the cmdline if it exists in the device tree.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
---
 arch/sh/Kconfig  |  8 ++++++++
 drivers/of/fdt.c | 11 +++++++++++
 2 files changed, 19 insertions(+)

--- a/arch/sh/Kconfig
+++ b/arch/sh/Kconfig
@@ -765,6 +765,14 @@ config CMDLINE
 	depends on CMDLINE_OVERWRITE || CMDLINE_EXTEND
 	default "console=ttySC1,115200"
 
+config CMDLINE_OVERRIDE
+    bool "Use alternative cmdline from device tree"
+    help
+      Some bootloaders may have uneditable bootargs. While CMDLINE_FORCE can
+      be used, this is not a good option for kernels that are shared across
+      devices. This setting enables using "chosen/cmdline-override" as the
+      cmdline if it exists in the device tree.
+
 endmenu
 
 menu "Bus options"
--- a/drivers/of/fdt.c
+++ b/drivers/of/fdt.c
@@ -1159,6 +1159,17 @@ int __init early_init_dt_scan_chosen(uns
 	if (p != NULL && l > 0)
 		strlcpy(data, p, min(l, COMMAND_LINE_SIZE));
 
+    /* CONFIG_CMDLINE_OVERRIDE is used to fallback to a different
+     * device tree option of chosen/bootargs-override. This is
+     * helpful on boards where u-boot sets bootargs, and is unable
+     * to be modified.
+     */
+#ifdef CONFIG_CMDLINE_OVERRIDE
+	p = of_get_flat_dt_prop(node, "bootargs-override", &l);
+	if (p != NULL && l > 0)
+		strlcpy(data, p, min((int)l, COMMAND_LINE_SIZE));
+#endif
+
 	/*
 	 * CONFIG_CMDLINE is meant to be a default in case nothing else
 	 * managed to set the command line, unless CONFIG_CMDLINE_FORCE
